### PR TITLE
Add support for tagging library datasets

### PR DESCRIPTION
**Pull request by abretaud**
This is the corresponding changes to the API change in https://github.com/galaxyproject/galaxy/pull/8530

### CODE REVIEW

File: bioblend/galaxy/libraries/__init__.py [hunk 1]
```diff
@@ -117,6 +117,9 @@ def update_library_dataset(self, dataset_id, **kwds):
         :type genome_build: str
         :param genome_build: Replace library dataset genome build (dbkey)
 
+        :type tags: list
+        :param tags: A list of tags to add to the dataset
```
Comment (nsoranzo) [inline bioblend/galaxy/libraries/__init__.py]: Can this also be used to remove tags?

File: bioblend/galaxy/libraries/__init__.py [hunk 2]
```diff
@@ -344,6 +347,8 @@ def _do_upload(self, library_id, **keywords):
         if keywords.get("link_data_only", None) and keywords['link_data_only'] != 'copy_files':
             payload["link_data_only"] = 'link_to_files'
         payload['tag_using_filenames'] = keywords.get('tag_using_filenames', True)
+        if keywords.get('tags', None) is not None:
```
Comment (nsoranzo) [inline bioblend/galaxy/libraries/__init__.py]:
```suggestion
        if keywords.get('tags'):
```

File: bioblend/galaxy/libraries/__init__.py (hunk 1 above)
Comment (abretaud) [reply bioblend/galaxy/libraries/__init__.py]: Actually it replaces existing tags with the ones given, I will change the docstring

Comment (abretaud) [reply bioblend/galaxy/libraries/__init__.py]: (and yes, I'll add tests!)

File: bioblend/_tests/TestGalaxyLibraries.py [hunk 3]
```diff
@@ -118,3 +118,62 @@ def test_dataset_permissions(self):
         self.assertEqual(set(_[1] for _ in ret_get['access_dataset_roles']), set(user_id_list_new))
         self.assertEqual(set(_[1] for _ in ret_get['modify_item_roles']), set(user_id_list_new))
         self.assertEqual(set(_[1] for _ in ret_get['manage_dataset_roles']), set(user_id_list_new))
+
+    @test_util.skip_unless_galaxy('release_19.09')
+    def test_upload_file_contents_with_tags(self):
+        with tempfile.NamedTemporaryFile(mode='w', prefix='bioblend_test_') as f:
```
Comment (nsoranzo) [inline bioblend/_tests/TestGalaxyLibraries.py]:
```suggestion
```

File: bioblend/_tests/TestGalaxyLibraries.py [hunk 4]
```diff
@@ -118,3 +118,62 @@ def test_dataset_permissions(self):
...
+            f.write(FOO_DATA)
```
Comment (nsoranzo) [inline bioblend/_tests/TestGalaxyLibraries.py]:
```suggestion
```

File: bioblend/_tests/TestGalaxyLibraries.py [hunk 5]
```diff
@@ -118,3 +118,62 @@ def test_dataset_permissions(self):
...
+            f.flush()
```
Comment (nsoranzo) [inline bioblend/_tests/TestGalaxyLibraries.py]:
```suggestion
```

File: bioblend/_tests/TestGalaxyLibraries.py [hunk 6]
```diff
@@ -118,3 +118,62 @@ def test_dataset_permissions(self):
...
+            datasets = self.gi.libraries.upload_file_contents(self.library['id'], FOO_DATA, tags=["foobar", "barfoo"])
```
Comment (nsoranzo) [inline bioblend/_tests/TestGalaxyLibraries.py]:
```suggestion
        datasets = self.gi.libraries.upload_file_contents(self.library['id'], FOO_DATA, tags=["foobar", "barfoo"])
```

File: bioblend/_tests/TestGalaxyLibraries.py [hunk 7]
```diff
@@ -118,3 +118,62 @@ def test_dataset_permissions(self):
...
+            dataset_show = self.gi.libraries.show_dataset(self.library['id'], datasets[0]['id'])
```
Comment (nsoranzo) [inline bioblend/_tests/TestGalaxyLibraries.py]:
```suggestion
        dataset_show = self.gi.libraries.show_dataset(self.library['id'], datasets[0]['id'])
```

File: bioblend/_tests/TestGalaxyLibraries.py [hunk 8]
```diff
@@ -118,3 +118,62 @@ def test_dataset_permissions(self):
...
+            self.assertEqual(dataset_show['tags'], 'name:foobar, name:barfoo')
```
Comment (nsoranzo) [inline bioblend/_tests/TestGalaxyLibraries.py]:
```suggestion
        self.assertEqual(dataset_show['tags'], 'name:foobar, name:barfoo')
```

File: bioblend/_tests/TestGalaxyLibraries.py [hunk 9]
```diff
@@ -118,3 +118,62 @@ def test_dataset_permissions(self):
...
+
+    @test_util.skip_unless_galaxy('release_19.09')
```
Comment (nsoranzo) [inline bioblend/_tests/TestGalaxyLibraries.py]: Given that all the upload methods just call `_do_upload()`, I think just testing tags with the `test_upload_file_contents_with_tags` test above is enough?
So `test_upload_file_local_path_with_tags` and `test_upload_from_galaxy_filesystem_with_tags` are not needed, IMO.

File: bioblend/_tests/TestGalaxyLibraries.py [hunk 10]
```diff
@@ -118,3 +118,62 @@ def test_dataset_permissions(self):
...
+                dataset = self.gi.libraries.wait_for_dataset(self.library['id'], dataset_dict['id'])
+                self.assertEqual(dataset['state'], 'ok')
+
+                dataset_show = self.gi.libraries.show_dataset(self.library['id'], dataset_dict['id'])
+                self.assertEqual(dataset_show['tags'], 'name:foobar, name:barfoo')
+        finally:
+            shutil.rmtree(tempdir)
+
+    @test_util.skip_unless_galaxy('release_19.09')
+    def test_upload_file_local_path_update_tags(self):
```
Comment (nsoranzo) [inline bioblend/_tests/TestGalaxyLibraries.py]:
```suggestion
    def test_update_dataset_tags(self):
```

File: bioblend/_tests/TestGalaxyLibraries.py [hunk 11]
```diff
@@ -118,3 +118,62 @@ def test_dataset_permissions(self):
...
+        with tempfile.NamedTemporaryFile(mode='w', prefix='bioblend_test_') as f:
+            f.write(FOO_DATA)
+            f.flush()
+            datasets = self.gi.libraries.upload_file_from_local_path(self.library['id'], f.name)
```
Comment (nsoranzo) [inline bioblend/_tests/TestGalaxyLibraries.py]: I'd use the faster `upload_file_contents` method as in `test_update_dataset()` above.

File: bioblend/_tests/TestGalaxyLibraries.py (hunk 9 above)
Comment (abretaud) [reply bioblend/_tests/TestGalaxyLibraries.py]: In fact tags are not added the same way between upload_from_galaxy_filesystem and the other methods. I wrote the 3 tests to understand why sometimes the tags were not added.
Now that it works for all, I removed test_upload_file_local_path_with_tags as it's clearly redundant, do you prefer to remove test_upload_from_galaxy_filesystem_with_tags also?

**Comment by abretaud** [comment]
I think the tests should pass now as https://github.com/galaxyproject/galaxy/pull/8530 was merged, maybe a rerun (or it can wait the 19.09 release)?

File: bioblend/_tests/TestGalaxyLibraries.py (hunk 9 above)
Comment (nsoranzo) [reply bioblend/_tests/TestGalaxyLibraries.py]: I think that tagging is now ensured for all code paths by https://github.com/galaxyproject/galaxy/pull/8530/files#diff-75e29f5ca09c0d8c3d9ef0286e1483dfR284-R286 , so I'd remove also `test_upload_from_galaxy_filesystem_with_tags` .

**Comment by nsoranzo** [comment]
Thanks @abretaud !

**Pull request merged**